llap-server/src/java/org/apache/hadoop/hive/llap/io/encoded/OrcEncodedDataReader.java

```diff
@@ -531,6 +531,15 @@ private void cleanupReaders() {
         // Ignore.
       }
     }
+
+    if (orcReader != null) {
```

Comment:
I might be wrong, but I think that is not right. Based on the code we reuse `orcReader` instance in multiple splits and create `stripeReader` per split which is closed in `cleanupReaders`.

see HIVE-21005: LLAP: Reading more stripes per-split leaks ZlibCodecs

Reply:
code is very strange. why do we create new `stripeReader` per stripe, can't we reuse?

Reply:
it looks like we close everything else in cleanupReaders, which is called from performDataRead
by closing this, are you suspecting a failure, or a perf degradation (meaning that this reader will be reopened any time if it's found to be null)
(let me check the lifecycle of the OrcEncodedDataReader in the meantime more thoroughly)

Reply:
maybe we should call `close` in a `call` method before we assign `orcReader = null`?

Reply:
also would be great if we manage to create Hive test with repro

Reply:
ok, looks like it's fine. 
we should close the reader in `getOrcTailForPath` as well. maybe use try-with-resources

Reply:
also please refactor to use `IOUtils.closeQuietly(orcReader)`

Reply:
yeah, I was thinking about making all the readers implement Closeable and use IOUtils.closeQuietly, so the question is where to stop refactoring the messy thing now :)

Reply:
:) I've added same comment

Reply:
stop gets invoked in LlapRecordReader.close, but since its another thread, we cannot close there so it has to be done in callInternal (same thread), and since callInternal is only invoked once per OrcEncodedDataReader(confirmed by Denys), changing code to do this in a try finally inside callInternal.

Reply:
not sure i get this, `orcReader` is created in LlapRecordReader constructor

Reply:
@harishjp, we should close in `stop`! What if we don't reach the call method, orcReader would be still open

Reply:
Yes you are right, that was one of the things I handled earlier with the AtomicInteger with state. I'll add that again then :-(.

Reply:
why do you need new state? there won't be 2 concurrent threads, just close the reader

Reply:
The implementation seems to be written such that stop() can be called when callInternal is executing, please look at how isStopped is being sent as AtomicBoolean and isStopped checks are in multiple places during callInternal, isn't that right? If yes then we need the new state, how else do you know that stop() is not being called when callInternal is executing?

Reply:
yes, looks like `isStopped` is used for thread interruption (i.e. `processStop`), but we are more interested when we can do the resource cleanup.